op-chain-ops/cmd/celo-migrate/main.go

```diff
@@ -0,0 +1,440 @@
+
+			// Build the L2 genesis block
+			l2Genesis, err := genesis.BuildL2Genesis(config, l1StartBlock)
+			if err != nil {
+				return fmt.Errorf("error creating l2 genesis: %w", err)
+			}
+
+			// Write changes to state to actual state database
+			cel2Header, err := ApplyMigrationChangesToDB(l2Genesis, dbPath, dbCache, dbHandles, !dryRun)
+			if err != nil {
+				return err
+			}
+			log.Info("Updated Cel2 state")
+
+			log.Info("Writing state diff", "file", outfileL2)
+			// Write genesis file to check created state
+			if err := jsonutil.WriteJSON(outfileL2, l2Genesis); err != nil {
+				return err
+			}
+
+			rollupConfig, err := config.RollupConfig(l1StartBlock, cel2Header.Hash(), cel2Header.Number.Uint64())
+			if err != nil {
+				return err
+			}
+			if err := rollupConfig.Check(); err != nil {
+				return fmt.Errorf("generated rollup config does not pass validation: %w", err)
+			}
+
+			log.Info("Writing rollup config", "file", outfileRollup)
+			if err := jsonutil.WriteJSON(outfileRollup, rollupConfig); err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	if err := app.Run(os.Args); err != nil {
+		log.Crit("error in migration", "err", err)
+	}
+	log.Info("Finished migration successfully!")
+}
+
+func ApplyMigrationChangesToDB(genesis *core.Genesis, dbPath string, dbCache int, dbHandles int, commit bool) (*types.Header, error) {
+	log.Info("Opening celo database", "dbCache", dbCache, "dbHandles", dbHandles, "dbPath", dbPath)
+	ldb, err := openCeloDb(dbPath, dbCache, dbHandles)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open DB: %w", err)
+	}
+	log.Info("Loaded Celo L1 DB", "db", ldb)
+
+	// Grab the hash of the tip of the legacy chain.
+	hash := rawdb.ReadHeadHeaderHash(ldb)
+	log.Info("Reading chain tip from database", "hash", hash)
+
+	// Grab the header number.
+	num := rawdb.ReadHeaderNumber(ldb, hash)
+	if num == nil {
+		return nil, fmt.Errorf("cannot find header number for %s", hash)
+	}
+	log.Info("Reading chain tip num from database", "number", num)
+
+	// Grab the full header.
+	header := rawdb.ReadHeader(ldb, hash, *num)
+	// trieRoot := header.Root
+	log.Info("Read header from database", "number", header)
+
+	// We need to update the chain config to set the correct hardforks.
+	genesisHash := rawdb.ReadCanonicalHash(ldb, 0)
+	cfg := rawdb.ReadChainConfig(ldb, genesisHash)
+	if cfg == nil {
+		log.Crit("chain config not found")
+	}
+	log.Info("Read chain config from database", "config", cfg)
+
+	dbFactory := func() (*state.StateDB, error) {
+		// Set up the backing store.
+		underlyingDB := state.NewDatabaseWithConfig(ldb, &trie.Config{
+			Preimages: true,
+		})
+
+		// Open up the state database.
+		db, err := state.New(header.Root, underlyingDB, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot open StateDB: %w", err)
+		}
+
+		return db, nil
+	}
+
+	db, err := dbFactory()
+	if err != nil {
+		return nil, fmt.Errorf("cannot create StateDB: %w", err)
+	}
```

Comment:
There doesn't seem to be a reason for this `dbFactory()`

```suggestion
	// Set up the backing store.
	underlyingDB := state.NewDatabaseWithConfig(ldb, &trie.Config{
		Preimages: true,
	})

	// Open up the state database.
	db, err := state.New(header.Root, underlyingDB, nil)
	if err != nil {
		return nil, fmt.Errorf("cannot open StateDB: %w", err)
	}
```

Reply:
I guess the reason for this was to make `underlyingDB` private and unusable by accident.